File: validation/poc-00/java-candidate/src/main/java/dev/vra/poc00/domain/InventoryBalance.java
package dev.vra.poc00.domain;

import java.util.Objects;

public record InventoryBalance(SkuId skuId, long onHand, long reserved, long version) {
    public InventoryBalance {
        Objects.requireNonNull(skuId, "skuId");
        if (onHand < 0 || reserved < 0 || reserved > onHand || version < 0)
            throw new IllegalArgumentException("Invalid inventory balance.");
    }
    public long available() { return onHand - reserved; }
    public static void requireQuantity(int quantity) {
        if (quantity <= 0) throw new DomainFailure.InvalidQuantity();
    }
    public InventoryBalance reserve(int quantity) {
        requireQuantity(quantity);
        if (quantity > available()) throw new DomainFailure.InsufficientStock();
        return new InventoryBalance(skuId, onHand, reserved + quantity, Math.incrementExact(version));
    }
}

File: validation/poc-00/kotlin-candidate/src/main/kotlin/dev/vra/poc00/kotlin/domain/InventoryBalance.kt
package dev.vra.poc00.kotlin.domain

data class InventoryBalance(
    val skuId: SkuId,
    val onHand: Long,
    val reserved: Long,
    val version: Long
) {
    init {
        require(onHand >= 0 && reserved >= 0 && reserved <= onHand && version >= 0) {
            "Invalid inventory balance."
        }
    }

    val available: Long get() = onHand - reserved

    fun reserve(quantity: Int): InventoryBalance {
        requireQuantity(quantity)
        if (quantity.toLong() > available) throw DomainFailure.InsufficientStock()
        return copy(reserved = reserved + quantity, version = Math.incrementExact(version))
    }

    companion object {
        fun requireQuantity(quantity: Int) {
            if (quantity <= 0) throw DomainFailure.InvalidQuantity()
        }
    }
}
